feat: Add support for accelerate-wrapped models in simple_evaluate()

DhruvaKashyap opened the pull request:
## Summary
Adds `HFLM_Accelerate` class to support models that are already wrapped with `accelerate`, enabling the use of `simple_evaluate()` with distributed and memory-optimized model instances.

## Problem
Currently, `simple_evaluate()` only accepts models as strings or base objects, making it incompatible with models wrapped using HuggingFace's `accelerate` library. This limits the library's utility for distributed evaluation scenarios. Addresses Issue #3312.

## Solution
- Added new `HFLM_Accelerate` class in `lm_eval/models/huggingface.py`
- Accepts pre-initialized `PreTrainedModel`, `PreTrainedTokenizerBase`, and `Accelerator` instances
- Inherits from `HFLM` but skips model loading since model is already initialized
- Maintains full compatibility with existing evaluation workflows

## Usage Example
```python
from accelerate import Accelerator
from transformers import AutoModel, AutoTokenizer
from lm_eval.models.huggingface import HFLM_Accelerate

accelerator = Accelerator()
model = AutoModel.from_pretrained("model_name")
tokenizer = AutoTokenizer.from_pretrained("model_name")

# Prepare model with accelerate
model = accelerator.prepare(model)

# Use with lm-evaluation-harness
hf_model = HFLM_Accelerate(
    pretrained=model,
    tokenizer=tokenizer, 
    accelerator=accelerator
)

DhruvaKashyap committed:
feat(models): add HFLM_Accelerate class for pre-loaded accelerate models

- Add HFLM_Accelerate class to support models already wrapped with accelerate
- Enables simple_evaluate() to work with distributed/accelerated model instances
- Inherits from HFLM but accepts pre-initialized PreTrainedModel and Accelerator
- Maintains compatibility with existing HFLM functionality

DhruvaKashyap committed:
Fix linting and type errors

DhruvaKashyap committed:
Merge branch 'EleutherAI:main' into feature/accelerate-model-support

DhruvaKashyap commented:
@baberabb Hi! Friendly bump on this PR. I’m happy to rebase, add tests, or make any necessary adjustments. Thanks!

DhruvaKashyap committed:
Merge branch 'EleutherAI:main' into feature/accelerate-model-support

StellaAthena commented:
I ran some tests and it seems like something is wrong. Firstly, I ran into a bug running this out of the box when using `accelerator.prepare(model)` with multi-GPU, the model gets wrapped in `DistributedDataParallel`, causing this error: `AttributeError: 'DistributedDataParallel' object has no attribute 'config'`.

The PR's `HFLM_Accelerate.__init__` accesses `self._model.config` directly at line 1622, but DDP wraps the model so config is at `model.module.config`. Without calling `accelerator.prepare()`, multi-GPU works. Each process loads its own model copy and processes a shard of the data. This achieves data parallelism through the accelerator's rank/world_size being correctly set.

I did not run into the issue @potpov describes. However this has quite poor scaling as shown in this table:

<img width="312" height="148" alt="Screenshot 2026-01-17 at 9 39 56 PM" src="https://github.com/user-attachments/assets/52b57380-10df-46de-9b12-d74941846e90" />

I haven't had time to investigate if this is expected or not yet.

DhruvaKashyap commented:
Hi @StellaAthena, thanks for taking the time to test this. 
You’re right about the `.config` bug; I seem to have missed this. Sorry for the confusion. 
I'll update the code with your fix. Would you recommend `accelerate.unwrap_model(self._model).config` or would using the `.module` interface be fine? The first adds some overhead, but may be a more stable interface to the underlying model.
Regarding the scaling, this wrapper tries to emulate the HFLM class with parallelize=True, and I would expect to see the same scaling as HFLM. But I'll try to get back to you on this.
Once again, thanks for taking the time.

DhruvaKashyap committed:
Merge branch 'EleutherAI:main' into feature/accelerate-model-support

DhruvaKashyap committed:
Fix direct access to model config

Fix a bug where the model config is directly accessed for a DDP model. Conif is now accessed through the module interface.

StellaAthena commented:
@DhruvaKashyap After some additional testing I've come to the conclusion that my original tests didn't run long enough and were heavily influenced by start-up time. One of the challenges here is that benchmarks have extremely different lengths and the amount of work needed to distribute them data parallel is sometimes not worth the gains in speed. In some tests it's faster than `parallelize=True` and in others it's slower.

I think it would make sense to run all three options (parallelize=True, launching using accelerate, and your PR) for some extremely large number of samples like 1 million and plot per-sample runtime as a function of the number of samples to see how these factors interact.

DhruvaKashyap commented:
Hi @StellaAthena,

My apologies for the very delayed response.

I've been trying to secure sufficient compute resources to run the experiment you requested, but unfortunately I haven't been able to do so yet.

That said, my understanding is that the primary motivation for this PR is orthogonal to efficiency. The ability to pass a model object directly, rather than requiring a HF string, is particularly valuable in accelerated environments for research workflows. It enables rapid evaluation of novel model architectures, sampling methods, and other experimental approaches that do not fit the standard `AutoModelForCausalLM` interface or correspond to a Hugging Face model string.

Examples include custom weight adapters, dynamically pruned networks, speculative decoding implementations, and other research-oriented modifications. These workflows are already supported by the current API in single-GPU settings; this PR simply extends analogous functionality to multi-GPU environments.

In my limited experiments with models under 10B parameters, I have observed performance improvements when using 4 GPUs across a set of standard evaluation tasks, including BoolQ, HellaSwag, WikiText perplexity, PIQA, ARC-E, ARC-C, and WinoGrande. However, I have not yet been able to perform the more comprehensive timing measurements you requested.

Given the scope and motivation of this PR, it would be helpful to understand whether those timing measurements are a strict requirement for moving forward with a merge, or whether the functionality itself is sufficient justification.

Thanks for your consideration.
